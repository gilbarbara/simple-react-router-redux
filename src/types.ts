import * as React from 'react';
import { Action, History, Location, LocationState } from 'history';
import { Middleware, Store } from 'redux';

const LOCATION_CHANGE = '@@router/LOCATION_CHANGE';

export interface LocationChangeAction<S = LocationState> {
  type: typeof LOCATION_CHANGE;
  payload: LocationChangePayload<S>;
}

export type LocationChangePayload<S = LocationState> = RouterState<S>;

export type Methods = 'go' | 'goBack' | 'goForward' | 'push' | 'replace';

export interface RouterAction {
  type: string;
  payload: {
    method: Methods;
    args: any[];
  };
}

export interface RouterLocation<S> extends Location<S> {
  query: Record<string, string>;
}

export type RouterMiddleware = <S = LocationState>(history: History<S>) => Middleware;

export interface RouterRootState<S = LocationState> {
  router: RouterState<S>;
}

export interface RouterState<S = LocationState> {
  location: RouterLocation<S>;
  action: Action;
}

export interface ConnectedRouterProps<S = LocationState> {
  children: React.FunctionComponent | React.ReactNode;
  history: History<S>;
  onLocationChanged: (location: Location<S>, action: Action) => LocationChangeAction<S>;
  store: Store;
}
