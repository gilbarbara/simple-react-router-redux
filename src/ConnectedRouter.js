import React from 'react';
import PropTypes from 'prop-types';
import { Router } from 'react-router';
import { connect } from 'react-redux';

import { onLocationChanged } from './actions';

class ConnectedRouter extends React.Component {
  constructor(props, context) {
    super(props);
    const { history, onLocationChanged: onChanged } = this.props;

    this.inTimeTravelling = false;
    this.unsubscribe = context.store.subscribe(() => {
      // Extract store's location
      const {
        pathname: pathnameInStore,
        search: searchInStore,
        hash: hashInStore,
      } = context.store.getState().router.location;
      // Extract history's location
      const {
        pathname: pathnameInHistory,
        search: searchInHistory,
        hash: hashInHistory,
      } = history.location;

      // If we do time travelling, the location in store is changed but location in history is not changed
      if (pathnameInHistory !== pathnameInStore || searchInHistory !== searchInStore || hashInHistory !== hashInStore) {
        this.inTimeTravelling = true;
        // Update history's location to match store's location
        history.push({
          pathname: pathnameInStore,
          search: searchInStore,
          hash: hashInStore,
        });
      }
    });

    const handleLocationChange = (location, action) => {
      if (!this.inTimeTravelling) {
        onChanged(location, action);
      }
      else {
        this.inTimeTravelling = false;
      }
    };

    this.unlisten = history.listen(handleLocationChange);
    handleLocationChange(history.location, history.action);
  }

  static propTypes = {
    children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    history: PropTypes.shape({
      action: PropTypes.string.isRequired,
      listen: PropTypes.func.isRequired,
      location: PropTypes.object.isRequired,
      push: PropTypes.func.isRequired,
    }).isRequired,
    onLocationChanged: PropTypes.func.isRequired,
  };

  static contextTypes = {
    store: PropTypes.shape({
      getState: PropTypes.func.isRequired,
      subscribe: PropTypes.func.isRequired,
    }).isRequired,
  };

  componentWillUnmount() {
    this.unlisten();
    this.unsubscribe();
  }

  render() {
    const { children, history } = this.props;

    return (
      <Router history={history}>
        {children}
      </Router>
    );
  }
}

const mapStateToProps = ({ router }) => ({
  action: router.action,
  location: router.location,
});

const mapDispatchToProps = dispatch => ({
  onLocationChanged: (location, action) => dispatch(onLocationChanged(location, action)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ConnectedRouter);
