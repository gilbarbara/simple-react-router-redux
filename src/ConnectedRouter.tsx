import * as React from 'react';
import { Router } from 'react-router-dom';
import { connect, ReactReduxContext, ReactReduxContextValue } from 'react-redux';
import { Dispatch } from 'redux';
import { Action, Location } from 'history';

import { onLocationChanged } from './actions';

import { ConnectedRouterProps, RouterLocation, RouterRootState } from './types';

class ConnectedRouter extends React.PureComponent<ConnectedRouterProps> {
  inTimeTravelling: boolean;
  unsubscribe: any;
  unlisten: any;

  constructor(props: ConnectedRouterProps) {
    super(props);
    const { history, onLocationChanged: onChanged, store } = this.props;

    this.inTimeTravelling = false;
    this.unsubscribe = store.subscribe(() => {
      // Extract store's location
      const {
        pathname: pathnameInStore,
        search: searchInStore,
        hash: hashInStore,
      } = store.getState().router.location;
      // Extract history's location
      const {
        pathname: pathnameInHistory,
        search: searchInHistory,
        hash: hashInHistory,
      } = history.location;

      // If we do time travelling, the location in store is changed but location in history is not changed
      /* istanbul ignore else */
      if (
        pathnameInHistory !== pathnameInStore ||
        searchInHistory !== searchInStore ||
        hashInHistory !== hashInStore
      ) {
        this.inTimeTravelling = true;
        // Update history's location to match store's location
        history.push({
          pathname: pathnameInStore,
          search: searchInStore,
          hash: hashInStore,
        });
      }
    });

    const handleLocationChange = (location: Location, action: Action) => {
      if (!this.inTimeTravelling) {
        onChanged(location, action);
      } else {
        this.inTimeTravelling = false;
      }
    };

    this.unlisten = history.listen(handleLocationChange);
    handleLocationChange(history.location, history.action);
  }

  componentWillUnmount() {
    this.unlisten();
    this.unsubscribe();
  }

  render() {
    const { children, history } = this.props;

    return <Router history={history}>{children}</Router>;
  }
}

const mapStateToProps = ({ router }: RouterRootState) => ({
  action: router.action,
  location: router.location,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  onLocationChanged: (location: RouterLocation<any>, action: Action) =>
    dispatch(onLocationChanged(location, action)),
});

const ConnectedRouterWithContext = (props: any) => {
  const { context } = props;
  const Context = context || ReactReduxContext;

  /* istanbul ignore next */
  if (Context == null) {
    throw new Error('Please upgrade to react-redux v6');
  }

  return (
    <Context.Consumer>
      {({ store }: ReactReduxContextValue) => <ConnectedRouter store={store} {...props} />}
    </Context.Consumer>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(ConnectedRouterWithContext);
