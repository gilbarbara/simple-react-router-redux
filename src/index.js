export * from './actions';
export ConnectedRouter from './ConnectedRouter';
export { default as routerMiddleware } from './middleware';
export { default as connectRouter } from './reducer';
