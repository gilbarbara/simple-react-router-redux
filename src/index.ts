export * from './actions';
export { default as ConnectedRouter } from './ConnectedRouter';
export { default as routerMiddleware } from './middleware';
export { default as connectRouter } from './reducer';

export * from './types';
