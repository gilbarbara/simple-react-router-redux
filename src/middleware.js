import { CALL_HISTORY_METHOD } from './actions';

/**
 * This middleware captures CALL_HISTORY_METHOD actions to redirect to the
 * provided history object. This will prevent these actions from reaching your
 * reducer or any middleware that comes after this one.
 */
const routerMiddleware = (history) => () => next => action => {
  if (action.type !== CALL_HISTORY_METHOD) {
    next(action);
    return;
  }

  const { payload: { method, args } } = action;
  history[method](...args);
};

export default routerMiddleware;
