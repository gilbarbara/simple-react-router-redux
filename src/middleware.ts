import { History } from 'history';
import { CALL_HISTORY_METHOD } from './actions';

import { RouterAction, RouterMiddleware } from './types';

/**
 * This middleware captures CALL_HISTORY_METHOD actions to redirect to the
 * provided history object. This will prevent these actions from reaching your
 * reducer or any middleware that comes after this one.
 */
const routerMiddleware: RouterMiddleware = (history: History) => () => next => (
  action: RouterAction,
) => {
  if (action.type !== CALL_HISTORY_METHOD) {
    next(action);

    return;
  }

  const { method, args } = action.payload;
  // @ts-ignore
  history[method](...args);
};

export default routerMiddleware;
