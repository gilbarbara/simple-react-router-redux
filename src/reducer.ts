import { Reducer } from 'redux';
import { History } from 'history';

import { LOCATION_CHANGE } from './actions';

export default (history: History): Reducer => {
  const initialRouterState = {
    location: history.location,
    action: history.action,
  };

  /*
   * This reducer will update the state with the most recent location history
   * has transitioned to.
   */
  return (state = initialRouterState, { type, payload }) => {
    if (type === LOCATION_CHANGE) {
      return {
        ...state,
        ...payload,
      };
    }

    return state;
  };
};
