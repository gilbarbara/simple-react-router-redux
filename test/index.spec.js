import * as ReactRouterRedux from '../src';

describe('simple-react-router-redux', () => {
  it('should ', () => {
    expect(ReactRouterRedux).toMatchSnapshot();
  });
});
