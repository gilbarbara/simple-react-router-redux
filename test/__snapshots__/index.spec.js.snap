// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple-react-router-redux should  1`] = `
Object {
  "CALL_HISTORY_METHOD": "@@router/CALL_HISTORY_METHOD",
  "ConnectedRouter": Object {
    "$$typeof": Symbol(react.memo),
    "WrappedComponent": [Function],
    "compare": null,
    "displayName": "Connect(ConnectedRouterWithContext)",
    "type": [Function],
  },
  "LOCATION_CHANGE": "@@router/LOCATION_CHANGE",
  "connectRouter": [Function],
  "go": [Function],
  "goBack": [Function],
  "goForward": [Function],
  "onLocationChanged": [Function],
  "push": [Function],
  "replace": [Function],
  "routerMiddleware": [Function],
}
`;
